lib/zephyr/src/zephyr/plan.py

```diff
@@ -0,0 +1,996 @@
+        for batch in process_chunk(all_chunk_specs[0]):
+            yield from batch
+    else:
+        futures = [exec_ctx.run(process_chunk, spec) for spec in all_chunk_specs]
+        yield from _merge_chunk_streams(exec_ctx, futures)
+
+
+def _sorted_merge_join(
+    left_stream: Iterable,
+    right_stream: Iterable,
+    left_key_fn: Callable,
+    right_key_fn: Callable,
+    combiner_fn: Callable,
+    join_type: str,
+) -> Iterator:
+    """Perform a sorted merge join between two streams.
+
+    Args:
+        left_stream: Iterator of items from left side
+        right_stream: Iterator of items from right side
+        left_key_fn: Function to extract key from left items
+        right_key_fn: Function to extract key from right items
+        combiner_fn: Function to combine matched items
+        join_type: "inner" or "left"
+
+    Yields:
+        Joined items according to join_type
+    """
+    # Materialize left stream and tag both streams
+    left_items = list(left_stream)
+    left_tagged = (("left", left_key_fn(item), item) for item in left_items)
+    right_tagged = (("right", right_key_fn(item), item) for item in right_stream)
+
+    # Merge both sorted streams by key
+    merged = heapq.merge(left_tagged, right_tagged, key=lambda x: x[1])
+
+    # Group by key and apply join logic
+    for _key, group in groupby(merged, key=lambda x: x[1]):
+        left_group: list[Any] = []
+        right_group: list[Any] = []
+        for side, _, item in group:
+            (left_group if side == "left" else right_group).append(item)
+
+        if join_type == "inner":
+            if left_group and right_group:
+                for left_item in left_group:
+                    for right_item in right_group:
+                        yield combiner_fn(left_item, right_item)
+        elif join_type == "left":
+            for left_item in left_group:
+                if right_group:
+                    for right_item in right_group:
+                        yield combiner_fn(left_item, right_item)
+                else:
+                    yield combiner_fn(left_item, None)
+
+
+@dataclass
+class StageContext:
+    """Context for executing a stage on a shard.
+
+    Encapsulates all the metadata and auxiliary data needed to process a shard.
+
+    Attributes:
+        shard: The shard data to process (iterable, typically a Shard object)
+        shard_idx: Index of this shard
+        total_shards: Total number of shards
+        chunk_size: Number of items per output chunk
+        aux_shards: Auxiliary shards for joins, keyed by op index
+        execution_context: Execution context for put/get/run/wait operations (for ForkChunks)
+    """
+
+    shard: Any  # Shard object (avoids circular import)
+    shard_idx: int
+    total_shards: int
+    chunk_size: int
+    aux_shards: dict[int, list[Any]] = field(default_factory=dict)
+    execution_context: Any = None  # ExecutionContext (avoids circular import)
```

Comment:
You may be able to guard the circular import behind `TYPE_CHECKING` and the use `"ExecutionContext"` (or `"JobContext"`)?

Reply:
sorry that was a claudeism, i fixed this and properly typed it.